### ISSUE TITLE
Improvements to the AudioHelper and AudioContainer API

### ISSUE THREAD

**Issue by aaclayton** [foundryvtt/foundryvtt#10092]
## Problems Addressed
This set of changes works to resolve the following problems or frictions with the V11 audio API.

- The separation of the `Sound` and `AudioContainer` API layers was unnecessarily complex.
- The `AudioContainer` did not have an ability to easily repeat playback of an `AudioBuffer` without looping since the buffer was disposed of upon completion.
- The `GainNode` of an `AudioContainer` did not survive repeated playback invocations preventing smooth volume transitions from being applied when playback of node is resumed.
- The `AudioContainer` did not have support for audio sprites which specify a `loopStart`, `loopEnd` (for looping) or `offset` and `duration` for one-time playback.
- The cache used for `AudioBuffer` was a naive `Map` rather than a more rigorous cache structure. This could incorrectly result in purging a buffer from the cache when multiple sound nodes use the same buffer.
- Timing of `window.setTimeout` is unreliable for audio events, numerous things can cause this to be wrong including loss of window focus.

## API Changes
- Combine the `Sound` and `AudioContainer` classes into a single `Sound` class which consolidates both functionalities into a simpler and more powerful interface.
- Allow `Sound#context` to be writable to assign a different `AudioContext` if desired. Context must be assigned before `Sound#play` is called.
- `Sound#load` now only loads the required `AudioBuffer` or `HTMLAudioElement`, but does not prepare the source node or connect it to downstream nodes. That workflow is deferred until `Sound#play` is called, allowing for buffered sounds to be easily replayed using a different `AudioBufferSourceNode`.
- Add support for an array of `SoundPlaybackOptions` which can be passed to the `Sound#play` method. This includes support for `loopStart` and `loopEnd` which can be used for audio sprites.
- Introduce an `AudioBufferCache` which more formally implements a LRU cache for `AudioBuffer`s. This replaces the naive `Map` that was previously used as `game.audio.buffers`. 
- Add the `CONFIG.debug.audio` flag which logs debugging information about audio playback if enabled.
- Expand `AudioContainer.LOAD_STATES` to the more general `Sound.STATES` which covers both loading and playback phases of the sound lifecycle.
- Introduce `AudioTimeout`, a specialized case of timeout management and an alternative to `window.setTimeout` which uses the web audio API to implement a very precise scheduling loop that is perfectly synchronized with other audio playback events.
- Improve `Sound#play`, `Sound#fade`, and `Sound#schedule` to use the `AudioTimeout` for managing asynchronous playback timing.
- Added `foundry.utils.EventEmitterMixin` which allows event-emission behavior to be added to any class.
- `AudioHelper#sounds` is changed from `Map<string, Sound>` to a `Map<string, WeakRef<Sound>>` for better memory management.
- Multiple calls to `Sound#load` are not managed behind a single shared promise which ensures that all requests are queued and handled once loading is complete.

## Breaking Changes
- `AudioHelper#getCache` is deprecated with backwards-compatibility in favor of `AudioHelper#buffers#get`
- `AudioHelper#setCache` is deprecated with backwards-compatibility in favor of `AudioHelper#buffers#set`
- `AudioHelper#updateCache` is deprecated without replacement
- `Sound#container` is deprecated without replacement because the `Sound` and `AudioContainer` classes are merged.
- `Sound#node` is renamed `Sound#sourceNode` as part of merger with `AudioContainer`
- `AudioContainer.LOAD_STATES` is removed in favor of `Sound.STATES`
- `AudioContainer.loadState` is deprecated in favor of `Sound#state`
- `Sound#on` is deprecated in favor of `Sound#addEventListener`
- `Sound#off` is deprecated in favor of `Sound#removeEventListener`
- `Sound#emit` is deprecated in favor of `Sound#dispatchEvent`

## Related Issues
- https://github.com/foundryvtt/foundryvtt/issues/7427
- https://github.com/foundryvtt/foundryvtt/issues/5632
- https://github.com/foundryvtt/foundryvtt/issues/10097
